### PR TITLE
Use Shared IRQ for stm32 ADC driver

### LINKED ISSUE #26874: STM32: Multiple ADCs using same interrupt not possible
**Is this request related to a missing driver support for a particular hardware platform, SoC or board? Please describe.**
Some STM32 chips have multiple ADCs. At the moment e.g. the STM32F407 ADC1 is the only available one out of three.

**Describe why you are asking for this support?**
On some of the larger chips (e.g. LFPQ144) the ADC3 has dedicated inputs I would like to use.

**If this is a new board or SoC, please state whether you are willing to maintain the Zephyr support for it if it is included in the main tree**
I am willing to implement this feature if I get some help.

I have attempted to add ADC2 and ADC3 to the device tree, but when multiple of them have the status "okay", I get the error 
```
gen_isr_tables.py: error: multiple registrations at table_index 18 for irq 18 (0x12)
Existing handler 0x801c283, new handler 0x801c283
Has IRQ_CONNECT or IRQ_DIRECT_CONNECT accidentally been invoked on the same irq multiple times?
```
Which is due to the fact that the three ADCs share interrupt 18 as a global ADC interrupt.

Addition is done in a .overlay:
```
/ {
    soc {
        adc2: adc@40012100 {
            compatible = "st,stm32-adc";
            reg = <0x40012100 0x100>;
            clocks = <&rcc STM32_CLOCK_BUS_APB2 0x00000100>;
            interrupts = <18 0>;
            status = "disabled";
            label = "ADC_2";
            #io-channel-cells = <1>;
        };
        adc3: adc@40012200 {
            compatible = "st,stm32-adc";
            reg = <0x40012200 0x100>;
            clocks = <&rcc STM32_CLOCK_BUS_APB2 0x00000100>;
            interrupts = <18 0>;
            status = "disabled";
            label = "ADC_3";
            #io-channel-cells = <1>;
        };
};
```
Could there be an option to nest the ADCs? This could be beneficial as they also share the ADC_Common register, which defines e.g. interleaving options.

### PR DESCRIPTION
**Pull request by mcscholtz**
I added the ability to use a shared IRQ for the stm32 ADC driver to allow devices with just a single ADC IRQ to use all channels.

Edit: This is my first PR, please let me know if I need to make any changes. Thanks!

Fixes #26874

### CODE REVIEW

**Comment by mcscholtz** [comment]
Looks like the failures are due to the configurable prescalar change. I will take a look later when I have time

**Comment by mcscholtz** [comment]
I was reading the contribution guidelines and I missed a few things, such as the commit signoff. I will fix that soon.

**Comment by mcscholtz** [comment]
I fixed the styling issues that caused the compliance checks to fail before. I also veried it with the checkpatch script before pushing it.

File: drivers/adc/Kconfig.stm32 [hunk 1]
```diff
@@ -3,10 +3,21 @@
...
+
+if SOC_SERIES_STM32F2X || SOC_SERIES_STM32F4X || SOC_SERIES_STM32F7X
+
+config ADC_STM32_SHARED_IRQS
+	bool "STM32 ADC shared interrupts"
+	default y
+	depends on ADC_STM32
+	help
+	  Enable the use of shared interrupts for famalies that only have a single interrupt for all ADC's
+endif
```
Comment (mbolivar-nordic) [inline drivers/adc/Kconfig.stm32:23]: Please add a newline to the end of the file.

File: drivers/adc/adc_stm32.c [hunk 2]
```diff
@@ -227,7 +236,6 @@ struct adc_stm32_data {
 	const struct device *dev;
 	uint16_t *buffer;
 	uint16_t *repeat_buffer;
-
```
Comment (mbolivar-nordic) [inline drivers/adc/adc_stm32.c:230]: We usually try not to include unrelated whitespace changes in commits which change functionality. I would recommend dropping this change or moving it to a separate commit.

File: drivers/adc/adc_stm32.c [hunk 3]
```diff
@@ -239,12 +247,17 @@ struct adc_stm32_data {
 
 struct adc_stm32_cfg {
 	ADC_TypeDef *base;
+	const uint16_t prescalar;
```
Comment (mbolivar-nordic) [inline drivers/adc/adc_stm32.c:250]: No need for `const` here as the entire structure is `const`.

File: drivers/adc/adc_stm32.c [hunk 4]
```diff
@@ -642,7 +687,7 @@ static int adc_stm32_check_acq_time(uint16_t acq_time)
 		return 0;
 	}
 
-	LOG_ERR("Conversion time not supportted.");
+	LOG_ERR("Conversion time not supported.");
```
Comment (mbolivar-nordic) [inline drivers/adc/adc_stm32.c:690]: You might want to split this into its own commit also.

File: dts/arm/st/f4/stm32f4.dtsi [hunk 5]
```diff
@@ -442,7 +443,7 @@
 
 		adc1: adc@40012000 {
 			compatible = "st,stm32-adc";
-			reg = <0x40012000 0x400>;
+			reg = <0x40012000 0x50>;
```
Comment (mbolivar-nordic) [inline dts/arm/st/f4/stm32f4.dtsi:446]: Is this really part of the same commit? I think it needs a justification.

**Review by mbolivar-nordic — CHANGES_REQUESTED** [review]
Thanks for your first patch! I have some general coding style comments. Please remember to follow the contribution guidelines by squashing or splitting commits as needed, without piling on a series of "fix this, fix that" commits to address review feedback.

@erwango or another STM32 maintainer will need to do the hardware specific review, but the binding change looks fine to me.

File: dts/arm/st/f4/stm32f4.dtsi (hunk 5 above)
Comment (mcscholtz) [reply dts/arm/st/f4/stm32f4.dtsi:446]: @mbolivar-nordic The ADC1 device currently holds the address space for ADC1, ADC2, ADC3. 

The address for all ADC's starts at 0x40012000 with a length of 0x400. The following is the addresses of the individual ADC's:
ADC1: 0x40012000 length of 0x50
ADC2: 0x40012100 length of 0x50
ADC3: 0x40012200 length of 0x50
Common: 0x40012300 length of 0xC (need to double check the length of this one)

Maybe someone that knows stm32 well can comment on how to handle it.

File: drivers/adc/adc_stm32.c (hunk 4 above)
Comment (mcscholtz) [reply drivers/adc/adc_stm32.c:690]: @mbolivar-nordic I can do that, would you recommend to just remove this and the other changes and force update the single comment. Then create new comments for the other changes, like the fix of this typo in this case?

File: dts/arm/st/f4/stm32f4.dtsi (hunk 5 above)
Comment (mbolivar-nordic) [reply dts/arm/st/f4/stm32f4.dtsi:446]: OK, then yes, I think this definitely needs a justification and likely a separate commit. 

> Maybe someone that knows stm32 well can comment on how to handle it.

Let's wait to see what the STM32 maintainers say.

File: drivers/adc/adc_stm32.c (hunk 4 above)
Comment (mbolivar-nordic) [reply drivers/adc/adc_stm32.c:690]: How people split up commits while rebasing is a personal decision, so I can't really offer you a "recommendation" in that sense. Beyond individual preference, tooling varies: for example, I use [magit](https://magit.vc/) instead of the command line.

However, one simple way to handle this since the PR contains a single commit is to put all its changes back into the working tree:

```
git checkout feature/stm32f407-adc
git reset HEAD~
```

Then you can split changes in a single file into their own commits using `git add -p`, e.g. for this particular comment:

```
git add -p drivers/adc/adc_stm32.c
< follow instructions to just stage this hunk>
git commit
<write a commit comment describing the typo fix>
```

Then keep adding additional commits for the other changes you want to make.

But again this sort of thing is subject to taste, and I would do it in a different way using my own favorite tools.

File: drivers/adc/adc_stm32.c [hunk 6]
```diff
@@ -968,6 +1048,34 @@ static const struct adc_driver_api api_stm32_driver_api = {
...
 
+#define STM32_ADC_INIT_SHARED(index)					\
+									\
+static const struct soc_gpio_pinctrl adc_pins_##index[] =		\
+	ST_STM32_DT_INST_PINCTRL(index, 0);				\
+									\
+static const struct adc_stm32_cfg adc_stm32_cfg_##index = {		\
+	.base = (ADC_TypeDef *)DT_INST_REG_ADDR(index),			\
+	.prescalar = DT_INST_PROP_OR(index, prescalar, 4),		\
+	.irq_cfg_func = adc_stm32_shared_irq_handler,			\
```
Comment (erwango) [inline drivers/adc/adc_stm32.c:1059]: This part of the struct is actually meant to hold the irq init function, not the irq handler.
In current case, it should then be NULL if we don't want to use it.
Alternatively, it could be set to `adc_stm32_cfg_func_isr` and use the `init_irq` flag inside this function to make sure irq init is only done once (which would remove need for the `adc_stm32_init` hack).

File: dts/arm/st/f4/stm32f4.dtsi (hunk 5 above)
Comment (erwango) [reply dts/arm/st/f4/stm32f4.dtsi:446]: I do agree with @mbolivar-nordic. Please split this in a dedicated commit.

File: drivers/adc/adc_stm32.c [hunk 7]
```diff
@@ -1005,4 +1114,8 @@ static void adc_stm32_cfg_func_##index(void)				\
 	irq_enable(DT_INST_IRQN(index));				\
 }
 
+#ifdef CONFIG_ADC_STM32_SHARED_IRQS
+DT_INST_FOREACH_STATUS_OKAY(STM32_ADC_INIT_SHARED)
+#else
 DT_INST_FOREACH_STATUS_OKAY(STM32_ADC_INIT)
+#endif
```
Comment (erwango) [inline drivers/adc/adc_stm32.c:1121]: Alternative otpion is to put the `#ifdef CONFIG_ADC_STM32_SHARED_IRQS` inside `STM32_ADC_INIT` definition, to minimize code changes

File: drivers/adc/adc_stm32.c [hunk 8]
```diff
@@ -681,6 +727,49 @@ static void adc_stm32_setup_speed(const struct device *dev, uint8_t id,
 #endif
 }
 
+#if defined(CONFIG_SOC_SERIES_STM32F2X) ||	\
+	defined(CONFIG_SOC_SERIES_STM32F4X) ||	\
+	defined(CONFIG_SOC_SERIES_STM32F7X)
+static uint32_t adc_stm32_get_prescalar(uint16_t prescalar)
```
Comment (erwango) [inline drivers/adc/adc_stm32.c:733]: Please put the prescalar changes into a dedicated commit.

File: drivers/adc/Kconfig.stm32 [hunk 9]
```diff
@@ -3,10 +3,21 @@
...
 	  Enable the driver implementation for the stm32xx ADC
+
+if SOC_SERIES_STM32F2X || SOC_SERIES_STM32F4X || SOC_SERIES_STM32F7X
+
+config ADC_STM32_SHARED_IRQS
+	bool "STM32 ADC shared interrupts"
+	default y
+	depends on ADC_STM32
+	help
+	  Enable the use of shared interrupts for famalies that only have a single interrupt for all ADC's
```
Comment (erwango) [inline drivers/adc/Kconfig.stm32:22]: s/famalies/families

**Review by erwango — CHANGES_REQUESTED** [review]
Thanks for this addition!
Some changes requested otherwise this looks fine.

Then, have you been able to test these changes ?

File: drivers/adc/adc_stm32.c [hunk 10]
```diff
@@ -984,7 +1038,16 @@ static const struct adc_stm32_cfg adc_stm32_cfg_##index = {		\
...
-};									\
+};
+#endif
+
+#define STM32_ADC_INIT(index)						\
+									\
+static const struct soc_gpio_pinctrl adc_pins_##index[] =		\
+	ST_STM32_DT_INST_PINCTRL(index, 0);				\
+									\
+	ADC_STM32_CONFIG_ISR(index)					\
```
Comment (erwango) [inline drivers/adc/adc_stm32.c:1049]: Unindent for readability.

File: drivers/adc/adc_stm32.c [hunk 11]
```diff
@@ -968,12 +973,61 @@ static const struct adc_driver_api api_stm32_driver_api = {
...
+	if (init_irq) {
+		init_irq = false;
+		IRQ_CONNECT(DT_INST_IRQN(0),
+			DT_INST_IRQ(0, priority),
+			adc_stm32_shared_irq_handler, NULL, 0);
+		irq_enable(DT_INST_IRQN(0));
+	}
+}
+
+#define ADC_STM32_CONFIG_ISR(index)					\
```
Comment (erwango) [inline drivers/adc/adc_stm32.c:1011]: I see why you used `ADC_STM32_CONFIG_ISR`, but this is a bit misleading.
Let's use `ADC_STM32_CONFIG` instead of `ADC_STM32_CONFIG_ISR`.

File: drivers/adc/adc_stm32.c [hunk 12]
```diff
@@ -968,12 +973,61 @@ static const struct adc_driver_api api_stm32_driver_api = {
...
+	if (adc_stm32_is_irq_active(adc_##index)) {				\
+		adc_stm32_isr(dev_##index);					\
+	}
+
+static void adc_stm32_shared_irq_handler(void)
+{
+	DT_INST_FOREACH_STATUS_OKAY(HANDLE_IRQS);
+}
+
+static void adc_stm32_cfg_func_isr(void)
```
Comment (erwango) [inline drivers/adc/adc_stm32.c:1000]: This funtion is not technically an isr, what about `adc_stm32_irq_init` ?

**Review by erwango — CHANGES_REQUESTED** [review]
Thanks for the update, few more comments.

**Comment by mcscholtz** [comment]
@erwango I made the changes you requested

**Review by erwango — APPROVED** [review]

File: drivers/adc/Kconfig.stm32 (hunk 1 above)
Comment (mcscholtz) [reply drivers/adc/Kconfig.stm32]: The compliance check failed after adding the new line at the end of the file

**Comment by mcscholtz** [comment]
@mbolivar-nordic Adding a line at the end of file drivers/adc/Kconfig.stm32 caused the compliance check to fail. Do you want me to remove it again?

**Comment by mbolivar-nordic** [comment]
> Adding a line at the end of file drivers/adc/Kconfig.stm32 caused the compliance check to fail. Do you want me to remove it again?

This can be a bit tricky, but there is a difference between "newline at end of file" and "empty line at end of file".

I'm asking for "newline at end of file", which means that the last byte in the file is a newline character. What is in there right now is "empty line at end of file", which is **two** newline characters.

**Comment by mcscholtz** [comment]
> I'm asking for "newline at end of file", which means that the last byte in the file is a newline character. What is in there right now is "empty line at end of file", which is **two** newline characters.

Okay, thanks for clarifying that. I have removed the extra new line

**Review by ABOSTM — APPROVED** [review]
LGTM

File: drivers/adc/adc_stm32.c [hunk 13]
```diff
@@ -982,12 +987,61 @@ static const struct adc_driver_api api_stm32_driver_api = {
...
+	}
+
+static void adc_stm32_shared_irq_handler(void)
+{
+	DT_INST_FOREACH_STATUS_OKAY(HANDLE_IRQS);
+}
+
+static void adc_stm32_irq_init(void)
+{
+	if (init_irq) {
```
[the remark this answers is not in the document]
Comment (erwango) [reply drivers/adc/adc_stm32.c:1016]: If we want this piece of code to be executed only once, we have to rely on a global variable, imo.

**Review by mbolivar-nordic — APPROVED** [review]

**Review by tejlmand — APPROVED** [review]

**Review by MaureenHelm — CHANGES_REQUESTED** [review]
> @mcscholtz You will have to rebase and make sure that you only have one ADC commit on top of the branch if you continue with this branch.

**Comment by carlescufi** [comment]
> > @mcscholtz You will have to rebase and make sure that you only have one ADC commit on top of the branch if you continue with this branch.

@MaureenHelm would you pick this PR up? Seems a shame not to merge it if the author is unresponsive.

**Comment by mcscholtz** [comment]
Closing. See https://github.com/zephyrproject-rtos/zephyr/pull/41497
